File: spark/src/main/python/passFunction.py
# -*- encoding: utf-8 -*-
"""
Created on 16:07 2019/4/9

@author: fancyChuan
@email: 1247375074@qq.com
@desc: 向spark传递函数 示例
"""
from pyspark import SparkConf, SparkContext


class SearchFunctionsWarn(object):
    """
    不推荐的写法，rdd.filter()传递函数的时候会把整个self都传递到集群上，可能self非常大
    """
    def __init__(self, query):
        self.query = query

    def isMatch(self, s):
        return self.query in s

    def getMatchesFunctionReference(self, rdd):
        return rdd.filter(self.isMathch)  # 有风险

    def getMatchesMemberReference(self, rdd):
        return rdd.filter(self.isMatch)  # 有风险


class SearchFunctionsRight(object):
    """
    推荐的写法，使用局部变量，把需要传递的内容赋值给局部变量，然后传递局部变量
    结论： 传递局部可序列化变量或者顶层函数始终是安全
    """
    def __init__(self, query):
        self.query = query

    def isMatch(self, s):
        return self.query in s

    def getMatchesFunctionReference(self, rdd):
        query = self.query
        return rdd.filter(lambda x: query in x)

    def getMatchesMemberReference(self, rdd):
        query = self.query
        return rdd.filter(lambda x: query in x)


if __name__ == "__main__":

    searchPython = SearchFunctionsWarn("Python")

    conf = SparkConf().setMaster("local").setAppName("passFunction")
    sc = SparkContext(conf=conf)
    linesRDD = sc.textFile("../resources/testfile.md")

    result = searchPython.getMatchesMemberReference(linesRDD)
    print(result.count())
    print(result.collect())

File: spark/src/main/java/learningSpark/rddProgram/PassFunction.java
package learningSpark.rddProgram;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.storage.StorageLevel;
import scala.Serializable;


/**
 * java版向Spark传递函数
 */
public class PassFunction implements Serializable {
    private JavaRDD<String> lines;

    public PassFunction() {
        SparkConf conf = new SparkConf().setMaster("local").setAppName("passFunction");
        JavaSparkContext sc = new JavaSparkContext(conf);
        this.lines = sc.textFile("E:\\JavaWorkshop\\bigdata-learn\\spark\\src\\main\\resources\\testfile.md");
        System.out.println("构造器初始化，共有行数：" + this.lines.count());
        System.out.println("持久化RDD...");
        this.lines.persist(StorageLevel.MEMORY_AND_DISK());
    }

    public PassFunction(JavaSparkContext sc) {
        this.lines = sc.textFile("E:\\JavaWorkshop\\bigdata-learn\\spark\\src\\main\\resources\\testfile.md");
        System.out.println("构造器初始化，共有行数：" + this.lines.count());
        System.out.println("持久化RDD...");
        this.lines.persist(StorageLevel.MEMORY_AND_DISK());
    }

    /**
     * 1. 使用匿名内部类进行函数传递
     */
    public void testAnonymousInnerClass() {
        JavaRDD<String> python = lines.filter(new Function<String, Boolean>() {
            @Override
            public Boolean call(String s) throws Exception {
                return s.contains("Python");
            }
        });
        System.out.println("使用匿名内部类：" + python.count());
    }

    /**
     * 2. 使用具名函数进行函数传递
     */
    public void testNamedClass() {
        JavaRDD<String> python = lines.filter(new ContainsPython());
        System.out.println("使用具名函数：" + python.count());
        // 传入关键词的具名类
        System.out.println("传入关键词的具名函数：" + lines.filter(new Contains("Spark")).count());
    }

    /**
     * 3. 使用Lambda表达式
     */
    public void testLambda() {
        JavaRDD<String> python = lines.filter(line -> line.contains("Python"));
        System.out.println("使用Lambda表达式：" + python.count());
    }
}

/**
 * 具名类
 */
class ContainsPython implements Function<String, Boolean>, Serializable {
    @Override
    public Boolean call(String s) throws Exception {
        return s.contains("Python");
    }
}

/**
 * 支持传关键词的具名类
 */
class Contains implements Function<String, Boolean>, Serializable {
    private String query;

    public Contains(String query) {
        this.query = query;
    }

    @Override
    public Boolean call(String s) throws Exception {
        return s.contains(query);
    }
}